src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -14,8 +15,237 @@
 
 namespace Microsoft.AspNetCore.DataProtection.KeyManagement
 {
+#if NETCOREAPP
     internal unsafe sealed class KeyRingBasedDataProtector : IDataProtector, IPersistedDataProtector
     {
+        public bool TryProtect(Span<byte> output, ReadOnlySpan<byte> plaintext, out int bytesWritten)
+        {
+            if (plaintext == null)
+            {
+                throw new ArgumentNullException(nameof(plaintext));
+            }
+
+            try
+            {
+                // Perform the encryption operation using the current default encryptor.
+                var currentKeyRing = _keyRingProvider.GetCurrentKeyRing();
+                var defaultKeyId = currentKeyRing.DefaultKeyId;
+                var defaultEncryptorInstance = currentKeyRing.DefaultAuthenticatedEncryptor;
+                CryptoUtil.Assert(defaultEncryptorInstance != null, "defaultEncryptorInstance != null");
+
+                if (_logger.IsDebugLevelEnabled())
+                {
+                    _logger.PerformingProtectOperationToKeyWithPurposes(defaultKeyId, JoinPurposesForLog(Purposes));
+                }
+
+                // We'll need to apply the default key id to the template if it hasn't already been applied.
+                // If the default key id has been updated since the last call to Protect, also write back the updated template.
+                var aad = _aadTemplate.GetAadForKey(defaultKeyId, isProtecting: true);
+
+                if (defaultEncryptorInstance.IsSpanEnabled())
+                {
+                    return defaultEncryptorInstance.TryEncrypt(output, plaintext, aad, out bytesWritten);
+                }
+
+                // We allocate a 20-byte pre-buffer so that we can inject the magic header and key id into the return value.
+                var retVal = defaultEncryptorInstance.Encrypt(
```

Comment:
I would expect this API to return false if the the encryptor did not support the span APIs required to implement this.

Reply:
This block of code is actually incorrect there's a test hole and the next update will fix the logic and add some tests, the code here is dropping the 20 byte pre-buffer.  The tests mock the interface and bypass the new span codepath, its unfortunate seems that I can't use mock with Span isn't useable as a type, so the new test will probably have to use the concrete implementations